File: ruby_dsl/action_pipeline/app/ad_execute/executors/participation/cpa_participate_executor.rb
module AdExecute
  module Executors
    class CpaParticipateExecutor < AdExecute::Executor
      include AdExecute::Concerns::LockManager
      include AdExecute::Concerns::TransactionManager

      def define_sequence
        action Actions::Macros::CommonPreValidate
        action Actions::Macros::Prepar`eParticipationContext

        with_lock(:ad_id, :uid) do
          action Actions::CheckDuplicateParticipation
          action Actions::ParticipateAdNetwork

          with_transaction do
            action Actions::CreateParticipateRecord
          end

          # DB 트랜잭션 밖이지만, 분산 락(Lock) 내부에서 실행하여 동시성 CAP 정합성 보장
          action Actions::IncrementAdReport
        end

        action Actions::Macros::GenerateClientResponse
        # CPA는 즉시 포스트백 없음
      end
    end
  end
end

File: kotlin_dsl/src/main/kotlin/ad_execute/executors/CpaParticipateExecutor.kt
package org.example.sample_kotlin.application.ad_execute.executors

import org.example.sample_kotlin.application.ad_execute.AdParticipateRequest
import org.example.sample_kotlin.application.ad_execute.Executor
import org.example.sample_kotlin.application.ad_execute.actions.*
import org.example.sample_kotlin.application.ad_execute.dsl.SequenceNode
import org.example.sample_kotlin.application.ad_execute.dsl.buildSequence
import org.springframework.stereotype.Component

/**
 * [ 비동기 보상형 (CPA, CPE) 광고 참여 파이프라인 ]
 * 특징: 유저가 렌딩을 타는 단계 (참여) 까지만 진행하고, 실제 적립(포스트백)은 요청하지 않고 종료함.
 */
@Component
class CpaParticipateExecutor : Executor() {
    
    override val sequence: List<SequenceNode> = buildSequence {
        action<ValidateRequestAction>()
        action<LoadAdResourceAction>()
        action<ValidateDeviceAction>()
        action<CheckParticipatableAdAction>()
        action<CheckBlockUserAction>()
        action<CheckPublisherStatusAction>()

        withLock("adType", "adId") {
            action<CheckDuplicateParticipationAction>()
            action<ParticipateAdNetworkAction>()
            
            withTransaction {
                action<CreateParticipateRecordAction>()
            }
        }
        action<GenerateLandingUrlAction>()
        // RequestPostbackAction 가 제외됨
    }

    override fun execute(request: AdParticipateRequest) {
        val context = hashMapOf(
            "uid" to request.uid,
            "adId" to request.adId,
            "advertisingId" to request.advertisingId,
            "adType" to request.adType
        )
        executeInternally(context)
    }
}

